src/dune_rules/warning_emit.mli

```diff
@@ -0,0 +1,13 @@
+(** Emit warnings that respect dune's conventions *)
+
+open Import
+
+type context =
+  | Build of Path.Build.t
+  | Source of Source_tree.Dir.t
+  | Source_dir of Path.Source.t
+  | Dune_project of Dune_project.t
+
+(** [emit w ctx f] will call [f] if [w] is enabled in [ctx]. [f] should
+    generate the warning corresponding to [w] *)
+val emit : Warning.t -> context -> (unit -> unit Memo.t) -> unit Memo.t
```

Comment:
The memoization is needed to look up what project we're in and if we're in a vendored directory. To raise without memoization, I suppose we could split the computation of if we're allowed to raise the warning from actually doing it. For example:

```
type context
val in_dir : Path.Source.t -> Warning.t -> context Memo.t
val emit : context -> (unit -> unit) -> unit
```

Reply:
I don't think that would make it any easier to raise in non-memo'd places in practice though. And I guess any part of the code that has enough information to create the context is probably going to be memo'd or at least it would probably be easy to spread memo to those parts of the code so I think it's fine to leave it for now and we can always split it up later if we have a use-case for non-memo'd `emit`.

Reply:
Emitting a context without consulting the context is definitely wrong though. Otherwise, we aren't going to respect the settings in the dune-project file for example.